internal/services/gitserver/repo-cleaner.go

```diff
@@ -0,0 +1,197 @@
+package gitserver
+
+import (
+	"context"
+	"io/ioutil"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"agola.io/agola/internal/services/config"
+	"agola.io/agola/internal/util"
+	"go.uber.org/zap"
+)
+
+type RepoCleaner struct {
+	dataDir          string
+	logger           *zap.SugaredLogger
+	scanTime         time.Duration
+	deleteBranchTime time.Duration
+}
+
+const (
+	DEFAULT_REPO_DELETE_BRANCH string = "720h"
+	DEFAULT_SCAN_TIME          string = "24h"
+)
+
+func NewRepoCleaner(c *config.Gitserver, l *zap.SugaredLogger) *RepoCleaner {
+	stDuration, err := time.ParseDuration(c.RepoCleaner.ScanTime)
+	if err != nil {
+		l.Error("repo-cleaner error:", err)
+		stDuration, _ = time.ParseDuration(DEFAULT_SCAN_TIME)
+	}
+
+	btDuration, err := time.ParseDuration(c.RepoCleaner.DeleteBranchAge)
+	if err != nil {
+		l.Error("repo-cleaner error:", err)
+		btDuration, _ = time.ParseDuration(DEFAULT_REPO_DELETE_BRANCH)
+	}
+
+	return &RepoCleaner{
+		dataDir:          c.DataDir,
+		logger:           l,
+		scanTime:         stDuration,
+		deleteBranchTime: btDuration,
+	}
+}
+
+func (s *RepoCleaner) Run(ctx context.Context) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				s.logger.Infof("repoCleaner exiting")
+
+				return
+			case <-time.After(s.scanTime):
+				if err := s.scanRepos(ctx); err != nil {
+					log.Errorf("err: %+v", err)
+				}
+			}
+		}
+	}()
+}
+
+func (s *RepoCleaner) scanRepos(ctx context.Context) error {
+	s.logger.Info("repoCleaner scanRepos start")
+
+	usersDir, err := ioutil.ReadDir(s.dataDir)
+	if err != nil {
+		return err
+	}
+
+	for _, u := range usersDir {
+		if u.IsDir() {
+			reposDir, _ := ioutil.ReadDir(s.dataDir + string(filepath.Separator) + u.Name())
+			for _, r := range reposDir {
+				if r.IsDir() {
+					s.scanRepo(ctx, s.dataDir+string(filepath.Separator)+u.Name()+string(filepath.Separator)+r.Name())
+				}
+			}
+		}
+	}
+
+	s.logger.Info("repoCleaner scanRepos end")
+
+	return nil
+}
+
+func (s *RepoCleaner) scanRepo(ctx context.Context, repoDir string) {
+	git := &util.Git{GitDir: repoDir}
+
+	branches, _ := s.getBranches(git, ctx)
+	deletedCount := 0
+	for _, b := range *branches {
+		date, _ := s.getLastCommitTime(ctx, git, "refs/heads/"+b)
+
+		if time.Since(*date) >= s.deleteBranchTime {
+			err := s.deleteBranch(ctx, git, b)
+			if err != nil {
+				s.logger.Error("repoCleaner error:", err)
+			} else {
+				deletedCount++
+			}
+		}
+	}
+
+	tags, _ := s.getTags(git, ctx)
+	for _, tag := range *tags {
+		date, _ := s.getLastCommitTime(ctx, git, "refs/tags/"+tag)
+
+		if time.Since(*date) >= s.deleteBranchTime {
+			err := s.deleteTag(ctx, git, tag)
+			if err != nil {
+				s.logger.Error("repoCleaner error:", err)
+			} else {
+				deletedCount++
+			}
+		}
+	}
+
```

Comment:
If you delete a tag or branch you're just removing a ref. So you should also prune dangling objects (see git prune) or they will remain until you don't remove the full repo.